src/core/urdb.jl

```diff
@@ -196,51 +195,34 @@ end
 
 
 """
-    parse_urdb_energy_costs(d::Dict, year::Int; time_steps_per_hour=1, bigM = 1.0e8)
+    parse_urdb_energy_costs(d::Dict, year::Int; time_steps_per_hour=1, bigM = 1.0e10)
 
 use URDB dict to return rates, energy_cost_vector, energy_tier_limits_kwh where:
     - rates is vector summary of rates within URDB (used for average rates when necessary)
     - energy_cost_vector is a vector of vectors with inner vectors for each energy rate tier,
         inner vectors are costs in each time step
-    - energy_tier_limits_kwh is a vector of upper kWh limits for each energy tier
+    - energy_tier_limits_kwh is a matrix of upper kWh limits for each energy tier, for each month
 """
-function parse_urdb_energy_costs(d::Dict, year::Int; time_steps_per_hour=1, bigM = 1.0e8)
+function parse_urdb_energy_costs(d::Dict, year::Int; time_steps_per_hour=1, bigM = 1.0e10)
     if length(d["energyratestructure"]) == 0
         throw(@error("No energyratestructure in URDB response."))
     end
-    # TODO check bigM (in multiple functions)
-    energy_tiers = Float64[]
-    for energy_rate in d["energyratestructure"]
-        append!(energy_tiers, length(energy_rate))
-    end
-    energy_tier_set = Set(energy_tiers)
-    if length(energy_tier_set) > 1
-        @warn "Energy periods contain different numbers of tiers, using limits of period with most tiers."
-    end
-    period_with_max_tiers = findall(energy_tiers .== maximum(energy_tiers))[1]
-    n_energy_tiers = Int(maximum(energy_tier_set))
-
-    energy_tier_limits_kwh = Float64[]
-
-    for energy_tier in d["energyratestructure"][period_with_max_tiers]
-        # energy_tier is a dictionary, eg. {'max': 1000, 'rate': 0.07531, 'adj': 0.0119, 'unit': 'kWh'}
-        energy_tier_max = get(energy_tier, "max", bigM)
-
-        if "rate" in keys(energy_tier) || "adj" in keys(energy_tier)  || "sell" in keys(energy_tier)
-            append!(energy_tier_limits_kwh, energy_tier_max)
-        end
-
-        if "unit" in keys(energy_tier) && string(energy_tier["unit"]) != "kWh"
-            throw(@error("URDB energy tiers have exotic units of " * energy_tier["unit"]))
-        end
-    end
-
+    scrub_urdb_tiers!(d["energyratestructure"])
+    n_energy_tiers = get_num_tiers(d["energyratestructure"])
     energy_cost_vector = Float64[]
     sell_vector = Float64[]
+    energy_tier_limits_kwh = Array{Float64}(undef, 12, n_energy_tiers)
 
     for tier in range(1, stop=n_energy_tiers)
 
         for month in range(1, stop=12)
+            period = Int(d["energyweekdayschedule"][month][1] + 1)
+            for tier in range(1, stop=n_energy_tiers)
+                # tiered energy schedules are assumed to be consistent for each month (i.e., the first hour can represent all 24 hours of the schedule).
```

Comment:
"the first hour can represent all 24 hours of the schedule" indicates to me that we couldn't model a TOU rate that varies by hour of day. Is that what this comment is intending to say?

Reply:
> So this means that if you have a rate with multiple periods within a month, and the tier maxes (kWh) vary between those periods, REopt will overwrite the tier max values for all periods that are not the hour 1 period, using hour 1 period maxes.
> 
> For example, period 1 here would be assumed to have a 1st tier max of 88 in May-Sept, and a 1st tier max of 100 in Nov (given the resulting energy_tier_limits matrix shown below); is that a correct interpretation? `energy_tier_limits_kwh: [88.0 1.0e10; 88.0 1.0e10; 88.0 1.0e10; 88.0 1.0e10; 88.0 1.0e10; 88.0 1.0e10; 88.0 1.0e10; 88.0 1.0e10; 88.0 1.0e10; 88.0 1.0e10; 100.0 1.0e10; 88.0 1.0e10]` <img alt="image" width="800" src="https://private-user-images.githubusercontent.com/16725585/337347047-d40c2ef2-06f8-41cf-8fcf-975bccd3e0b0.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTgxMTU3OTMsIm5iZiI6MTcxODExNTQ5MywicGF0aCI6Ii8xNjcyNTU4NS8zMzczNDcwNDctZDQwYzJlZjItMDZmOC00MWNmLThmY2YtOTc1YmNjZDNlMGIwLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA2MTElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNjExVDE0MTgxM1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTYxZmIxMDYzOTcxOWZjZDkzMTI1NGEwZjM3MWY4YjAzNmY4ZTkzMjYwZTkyZDQ0MjliMTg0YjhmYTAxNTliMmEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.C-7-PoWVps4pq3X4h1BoGs-tlWW3YofFPjw8w25PLuE">
> 
> Is it accurate to say this is still not fully capturing the intended rate, but is one step better than what we did before, which was assume that all periods[tiers] had the same max limits?

That's correct if we are talking about monthly rates and limits, but the time-of-use "ratchets" allow for custom definition of periods and so this setup likely assigns a separate "ratchet" that consists of thirty hourly periods from 12-1am and another for all the other hours in November, each of which has its own separate tiers and limits.  I'd expect to see 18 time-of-use periods in this setup, each of which has two tiers.

So, to that end: is the above matrix what's actually produced when you insert this custom output or do you get the 2x18 matrix?  One thing I did not touch as part of this PR is the logic for when custom time-of-use vs. monthly tiers are generated.  We can revisit that, but it's likely best reserved for an issue at this point.

Reply:
piggybacking on this comment: what I said above applies to demand charges but **not** energy rates, so this is a missing feature as compared to everything that the URDB rate builder allows for.